actionview/test/template/form_options_helper_test.rb

```diff
@@ -715,6 +715,19 @@ def test_select_under_fields_for_with_block_without_options
     )
   end
 
+  def test_select_under_fields_for_with_block_returning_a_non_string_value
+    @post = Post.new
+
+    output_buffer = fields_for :post, @post do |f|
+      concat(f.select(:category) { [] })
```

Comment:
What's the expectation here, that it'd return an array of choices, rather than option tags? That seems new behavior, can we test that it works like that?

Reply:
Great question!

No, the expectation remains the same: the block either outputs several option tags that get captured into the output buffer or returns a string that includes the option tags.

The problem is that when the helper is used as suggested in the docs to output several option tags within a call to `each`, and `available_campaigns` is empty, no option tags will be output, and the block will effectively return `[]` which will cause `@choices` in `ActionView::Helpers::Tags::Select#initialize` to be set to `nil`.

```
select(report, :campaign_ids) do
  available_campaigns.each do |c|
    tag.option(c.name, value: c.id, data: { tags: c.tags.to_json })
  end
end
```

This will in turn cause the following error in `options_for_select`:

```
NoMethodError: undefined method `map' for nil
    lib/action_view/helpers/form_options_helper.rb:365:in `options_for_select'
    lib/action_view/helpers/tags/select.rb:28:in `render'
    lib/action_view/helpers/form_options_helper.rb:160:in `select'
    lib/action_view/helpers/form_options_helper.rb:849:in `select'
    test/template/form_options_helper_test.rb:722:in `block in test_select_under_fields_for_with_block_returning_a_non_string_value'
    lib/action_view/helpers/capture_helper.rb:50:in `block in capture'
    lib/action_view/buffers.rb:75:in `capture'
    lib/action_view/helpers/capture_helper.rb:50:in `capture'
    lib/action_view/helpers/form_helper.rb:1088:in `fields'
    lib/action_view/helpers/form_helper.rb:1031:in `fields_for'
    test/template/form_options_helper_test.rb:721:in `test_select_under_fields_for_with_block_returning_a_non_string_value'
```

Would changing this line to something like this explain the situation better?

```suggestion
      concat(f.select(:category) { [].each { |title| content_tag(:option, title) })
```

Reply:
I'm surprised that the output of the block becomes equivalent to whatever `.each` was invoked on, I would expect the `capture` helper to return an empty string when no output is generated. Maybe that's the underlying bug that should be addressed?

Reply:
Is the underlying problem that `capture` should uniformly return a string, even when the block produces no output, or output that is not a string? The behavior now is...

```
capture { "a" }=> "a"
capture { 1 } => nil
capture { [] } => nil
```

should it instead be:

```
capture { "a" } => "a"
capture { 1 } => ""
capture { [] } => ""
```

Reply:
@meagar I'm not sure if there is any value in capture returning a nil in these cases. Would be cool to see a PR trying to change it, though, to see if CI passes!

Reply:
@calleluks I'll have a look